octobot_services/service_feeds/abstract_service_feed.py

```diff
@@ -0,0 +1,148 @@
+#  Copyright (c) Drakkar-Software, All rights reserved.
+#
+#  This library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public
+#  License as published by the Free Software Foundation; either
+#  version 3.0 of the License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+
+from abc import *
+import threading
+
+from octobot_channels.channels.channel import get_chan, set_chan
+
+from octobot_commons.asyncio_tools import run_coroutine_in_asyncio_loop
+
+
+# ****** Unique dispatcher side ******
+from octobot_services.abstract_service_user import AbstractServiceUser
+from octobot_services.channel.abstract_service_feed import AbstractServiceFeedChannelProducer
+
+
+class AbstractServiceFeed(AbstractServiceUser, threading.Thread, AbstractServiceFeedChannelProducer):
+    __metaclass__ = ABCMeta
+
+    # Override FEED_CHANNEL with a dedicated channel
+    FEED_CHANNEL = None
+
+    _SLEEPING_TIME_BEFORE_RECONNECT_ATTEMPT_SEC = 10
+    DELAY_BETWEEN_STREAMS_QUERIES = 5
+    REQUIRED_SERVICE_ERROR_MESSAGE = "Required services are not ready, service feed can't start"
+
+    def __init__(self, config, main_async_loop):
+        threading.Thread.__init__(self)
+        AbstractServiceUser.__init__(self, config)
+        AbstractServiceFeedChannelProducer.__init__(self, set_chan(self.FEED_CHANNEL(), None))
+        self.feed_config = {}
+        self.main_async_loop = main_async_loop
+        self.service = None
+
+    # Override update_feed_config if any need in the extending feed
+    def update_feed_config(self, config):
+        pass
+
+    # Override this method if the service feed implementation is using a dispatcher handled in the service layer
+    # (ie: TelegramServiceFeed)
+    @staticmethod
+    def _get_service_layer_service_feed() -> object:
+        return None
+
+    # Override this method to specify the feed reception process
+    @abstractmethod
+    def _start_service_feed(self):
+        raise NotImplementedError("start_dispatcher not implemented")
+
+    @abstractmethod
+    def _something_to_watch(self):
+        raise NotImplementedError("_something_to_watch not implemented")
+
+    @abstractmethod
+    def _initialize(self):
+        raise NotImplementedError("_initialize not implemented")
+
+    def _init_chanel(self):
+        channel = get_chan(self.FEED_CHANNEL.get_name())
+        run_coroutine_in_asyncio_loop(channel.register_producer(self),
+                                      self.main_async_loop)
+
+    # Call _notify_consumers to send data to consumers
+    def _notify_consumers(self, data):
```

Comment:
Why `_notify_consumers` is not async ?

Reply:
For now all 3 service feeds are using threads (direct ones for reddit and twitter and a module provided one for telegram). Therefore they are not in the main async loop. 
We can either make ```_notify_consumers``` async and call it with a ```run_coroutine_in_asyncio_loop``` wrapper for the case when we will have a main loop async service feed or leave it like this (which is in my opinion simpler) and make the change when we will have such an async service feed

Reply:
OK :+1: